engine/orders.go

```diff
@@ -17,47 +17,121 @@ import (
 var (
 	OrderManagerDelay      = time.Second * 10
 	ErrOrdersAlreadyExists = errors.New("order already exists")
+	ErrOrderNotFound       = errors.New("order does not exist")
 )
 
-func (o *orderStore) Get() map[string][]order.Detail {
+// get returns all orders for all exchanges
+// should not be exported as it can have large impact if used improperly
+func (o *orderStore) get() map[string][]*order.Detail {
 	o.m.Lock()
 	defer o.m.Unlock()
 	return o.Orders
 }
 
+// GetByExchangeAndID returns a specific order
+func (o *orderStore) GetByExchangeAndID(exchange, id string) (*order.Detail, error) {
+	o.m.Lock()
+	defer o.m.Unlock()
+	r, ok := o.Orders[exchange]
+	if !ok {
+		return nil, ErrExchangeNotFound
+	}
+
+	for x := range r {
+		if r[x].ID == id {
+			return r[x], nil
+		}
+	}
+	return nil, ErrOrderNotFound
+}
+
+// GetByExchangeAndID returns a specific order
+func (o *orderStore) GetByExchange(exchange string) ([]*order.Detail, error) {
+	o.m.Lock()
+	r, ok := o.Orders[exchange]
+	if !ok {
+		o.m.Unlock()
+		return nil, ErrExchangeNotFound
+	}
+	o.m.Unlock()
+	return r, nil
+}
+
+// GetByInternalOrderID will search all orders for our internal orderID
+// and return the order
+func (o *orderStore) GetByInternalOrderID(internalOrderID string) (*order.Detail, error) {
+	o.m.Lock()
+	for _, v := range o.Orders {
+		for x := range v {
+			if v[x].InternalOrderID == internalOrderID {
+				o.m.Unlock()
+				return v[x], nil
+			}
+		}
+	}
+	o.m.Unlock()
+	return nil, ErrOrderNotFound
+}
+
 func (o *orderStore) exists(order *order.Detail) bool {
+	if order == nil {
+		return false
+	}
+	o.m.Lock()
 	r, ok := o.Orders[order.Exchange]
 	if !ok {
+		o.m.Unlock()
 		return false
 	}
 
 	for x := range r {
 		if r[x].ID == order.ID {
+			o.m.Unlock()
 			return true
 		}
 	}
-
+	o.m.Unlock()
 	return false
 }
 
+// Adds an order to the orderStore for tracking the lifecycle
 func (o *orderStore) Add(order *order.Detail) error {
-	o.m.Lock()
-	defer o.m.Unlock()
-
+	if order == nil {
+		return errors.New("order store: Order is nil")
+	}
+	exch := GetExchangeByName(order.Exchange)
+	if exch == nil {
+		return errors.New("unable to get exchange by name")
+	}
 	if o.exists(order) {
 		return ErrOrdersAlreadyExists
 	}
-
+	// Untracked websocket orders will not have internalIDs yet
+	if order.InternalOrderID == "" {
+		id, err := uuid.NewV4()
+		if err != nil {
```

Comment:
At this point order.InternalOrderID wont be set to anything is there ever a case where we would need to set these to some generic known id so they are not blank later?

Reply:
This one is weird because its dependent on an external library. It failing will impact many things, but I don't think it should stop the logging of an order. If it ends up erroring, later down the line, I'd want it to still have a GUID there for consistency, but if its GUID gen is erroring, I can't really put the same code in at a later stage. I'm not sure what to do here

Reply:
We can add our own internal GUID generator in the case that this happens, even though it should theoretically be close to nil

Reply:
Yeah not sure on a clean fix

I was more just thinking if an order is added with no internal id due to uuid in the very rare case erroring GetByInternalOrderID() wont work 

So in that case is it work not adding the order? how important is it that intenralorderid is filled out correctly?

Reply:
Right now, its actually not very important. The order manager is not storing orders in the database, so when you reload the application, you would end up with the same order having a different GUID. So I think our own GUID handling as well as order storage will be in a future PR

Reply:
Happy with that good to resolve